Servers/DHCPClient/DHCPv4.h

```diff
@@ -0,0 +1,348 @@
+            builder.append('\n');
+        }
+        return builder.build();
+    }
+
+    struct DHCPOptionValue {
+        u8 length;
+        const void* value;
+    };
+
+    HashMap<DHCPOptions, DHCPOptionValue> options;
+};
+
+class [[gnu::packed]] DHCPv4Packet
+{
+public:
+    u8 op() const { return m_op; }
+    void set_op(DHCPv4Ops op) { m_op = (u8)op; }
+
+    u8 htype() const { return m_htype; }
+    void set_htype(u8 htype) { m_htype = htype; }
+
+    u8 hlen() const { return m_hlen; }
+    void set_hlen(u8 hlen) { m_hlen = hlen; }
+
+    u8 hops() const { return m_hops; }
+    void set_hops(u8 hops) { m_hops = hops; }
+
+    u32 xid() const { return m_xid; }
+    void set_xid(u32 xid) { m_xid = xid; }
+
+    u16 secs() const { return m_secs; }
+    void set_secs(u16 secs) { m_secs = secs; }
+
+    u16 flags() const { return m_flags; }
+    void set_flags(DHCPv4Flags flags) { m_flags = (u16)flags; }
+
+    const IPv4Address& ciaddr() const { return m_ciaddr; }
+    const IPv4Address& yiaddr() const { return m_yiaddr; }
+    const IPv4Address& siaddr() const { return m_siaddr; }
+    const IPv4Address& giaddr() const { return m_giaddr; }
+
+    IPv4Address& ciaddr() { return m_ciaddr; }
+    IPv4Address& yiaddr() { return m_yiaddr; }
+    IPv4Address& siaddr() { return m_siaddr; }
+    IPv4Address& giaddr() { return m_giaddr; }
+
+    u8* options() { return m_options; }
+    ParsedDHCPv4Options parse_options() const;
+
+    const MACAddress& chaddr() const { return *(const MACAddress*)&m_chaddr[0]; }
+    void set_chaddr(const MACAddress& mac) { *(MACAddress*)&m_chaddr[0] = mac; }
+
+    StringView sname() const { return { (const char*)&m_sname[0] }; }
+    StringView file() const { return { (const char*)&m_file[0] }; }
+
+private:
+    NetworkOrdered<u8> m_op;
+    NetworkOrdered<u8> m_htype;
+    NetworkOrdered<u8> m_hlen;
+    NetworkOrdered<u8> m_hops;
+    NetworkOrdered<u32> m_xid;
+    NetworkOrdered<u16> m_secs;
+    NetworkOrdered<u16> m_flags;
+    IPv4Address m_ciaddr;
+    IPv4Address m_yiaddr;
+    IPv4Address m_siaddr;
+    IPv4Address m_giaddr;
+    u8 m_chaddr[16]; // 10 bytes of padding at the end
+    u8 m_sname[64] { 0 };
+    u8 m_file[128] { 0 };
+    u8 m_options[312] { 0 }; // variable, less than 312 bytes
+};
+
+class DHCPv4PacketBuilder {
+public:
+    DHCPv4PacketBuilder()
+        : m_buffer(ByteBuffer::create_zeroed(sizeof(DHCPv4Packet)))
+    {
+        auto* options = peek().options();
+        options[0] = 99;
+        options[1] = 130;
+        options[2] = 83;
+        options[3] = 99;
+    }
+
+    void add_option(DHCPOptions option, u8 length, const void* data)
+    {
+        ASSERT(m_can_add);
+        auto* options = peek().options();
+        options[next_option_offset++] = (u8)option;
+        __builtin_memcpy(options + next_option_offset, &length, 1);
+        next_option_offset++;
+        __builtin_memcpy(options + next_option_offset, data, length);
```

Comment:
You can just use regular `memcpy()` here. The use of `__builtin_memcpy` in `AK` headers is a trick to avoid pulling in LibC headers. :)